packages/components/src/navigator/types.ts

```diff
@@ -12,16 +12,27 @@ export type NavigatorLocation = NavigateOptions & {
 	isBack?: boolean;
 	path?: string;
 	hasRestoredFocus?: boolean;
+	restoreFocusTo?: string;
 };
 
 export type NavigatorContext = {
 	location: NavigatorLocation;
+	params: any;
```

Comment:
I've tried to change this but failed due to the types of the path-to-regexp library we're using. Maybe you could give it a try if you have some time.

Reply:
I had a deeper look at `path-to-regexp`, and it turns out that  `params` can also return arrays on top if simple strings.

For example, calling `patternMatch( '/some/basic/route', [ { id: 'route', path: '/:test+' } ] )` would return 

```
Object {
  "id": "route",
  "params": Object {
    "test": Array [
      "some",
      "basic",
      "route",
    ],
  },
}
```

The simplest solution that comes to mind is to type `params` as `Record< string, string | string[] >` and let the consumer of `Navigator` and `NavigatorScreen` handle the `params` accordingly (ie. checking if it's an array or a string).

While playing around with these types, I also tried putting the matching screen ID and the params under the same object in the `NavigatorContext`

<details>

<summary>Here's what I came up with, which seems to work well:</summary>

```diff
diff --git a/packages/components/src/navigator/navigator-provider/component.tsx b/packages/components/src/navigator/navigator-provider/component.tsx
index ab381b42d7..c6a61ce6f0 100644
--- a/packages/components/src/navigator/navigator-provider/component.tsx
+++ b/packages/components/src/navigator/navigator-provider/component.tsx
@@ -70,7 +70,7 @@ function UnconnectedNavigatorProvider(
 	] );
 	const [ screens, dispatch ] = useReducer( screensReducer, [] );
 	const currentMatch = useRef< MatchedPath >();
-	const matchedPath = useMemo( () => {
+	const match = useMemo( () => {
 		const resolvePath = ( path: string ) => {
 			const newMatch = patternMatch( path, screens );
 
@@ -170,14 +170,13 @@ function UnconnectedNavigatorProvider(
 				...locationHistory[ locationHistory.length - 1 ],
 				isInitial: locationHistory.length === 1,
 			},
-			params: matchedPath ? matchedPath.params : {},
-			match: matchedPath ? matchedPath.id : undefined,
+			match,
 			goTo,
 			goBack,
 			addScreen,
 			removeScreen,
 		} ),
-		[ locationHistory, matchedPath, goTo, goBack, addScreen, removeScreen ]
+		[ locationHistory, match, goTo, goBack, addScreen, removeScreen ]
 	);
 
 	const cx = useCx();
diff --git a/packages/components/src/navigator/navigator-screen/component.tsx b/packages/components/src/navigator/navigator-screen/component.tsx
index 15cec834d8..6d9dec5c9e 100644
--- a/packages/components/src/navigator/navigator-screen/component.tsx
+++ b/packages/components/src/navigator/navigator-screen/component.tsx
@@ -59,7 +59,7 @@ function UnconnectedNavigatorScreen(
 	const prefersReducedMotion = useReducedMotion();
 	const { location, match, addScreen, removeScreen } =
 		useContext( NavigatorContext );
-	const isMatch = match === screenId;
+	const isMatch = match?.id === screenId;
 	const wrapperRef = useRef< HTMLDivElement >( null );
 
 	useEffect( () => {
diff --git a/packages/components/src/navigator/test/router.ts b/packages/components/src/navigator/test/router.ts
index 77b04e38c9..7c60b846a4 100644
--- a/packages/components/src/navigator/test/router.ts
+++ b/packages/components/src/navigator/test/router.ts
@@ -37,4 +37,14 @@ describe( 'patternMatch', () => {
 		] );
 		expect( result ).toEqual( { id: 'route', params: {} } );
 	} );
+
+	it( 'should return an array of matches for the same param', () => {
+		const result = patternMatch( '/some/basic/route', [
+			{ id: 'route', path: '/:test+' },
+		] );
+		expect( result ).toEqual( {
+			id: 'route',
+			params: { test: [ 'some', 'basic', 'route' ] },
+		} );
+	} );
 } );
diff --git a/packages/components/src/navigator/types.ts b/packages/components/src/navigator/types.ts
index 6833cffea4..fb8b5eeea8 100644
--- a/packages/components/src/navigator/types.ts
+++ b/packages/components/src/navigator/types.ts
@@ -8,6 +8,8 @@ import type { ReactNode } from 'react';
  */
 import type { ButtonAsButtonProps } from '../button/types';
 
+export type MatchParams = Record< string, string | string[] >;
+
 type NavigateOptions = {
 	focusTargetSelector?: string;
 };
@@ -22,8 +24,10 @@ export type NavigatorLocation = NavigateOptions & {
 // Returned by the `useNavigator` hook.
 export type Navigator = {
 	location: NavigatorLocation;
-	params: any;
-	match?: string;
+	match?: {
+		id: string;
+		params?: MatchParams;
+	};
 	goTo: ( path: string, options?: NavigateOptions ) => void;
 	goBack: () => void;
 };
diff --git a/packages/components/src/navigator/utils/router.ts b/packages/components/src/navigator/utils/router.ts
index 4dd5263fd7..0aad26e582 100644
--- a/packages/components/src/navigator/utils/router.ts
+++ b/packages/components/src/navigator/utils/router.ts
@@ -6,11 +6,11 @@ import { match } from 'path-to-regexp';
 /**
  * Internal dependencies
  */
-import type { NavigatorScreen } from '../types';
+import type { NavigatorScreen, MatchParams } from '../types';
 
 export function patternMatch( path: string, screens: NavigatorScreen[] ) {
 	for ( const screen of screens ) {
-		const matchingFunction = match( screen.path, {
+		const matchingFunction = match< MatchParams >( screen.path, {
 			decode: decodeURIComponent,
 		} );
 		const matched = matchingFunction( path );
```
</details>

Reply:
After some thoughts, I think I'd rather keep "match" (the id) as a private API and only expose params in useNavigator because the id is generated internally, so I'll just keep the id separate but not expose it publicly.

Reply:
That's a good point, I agree about keeping the matching ID as private (not exposed in `useNavigator`)